Add separate OME-TIFF tests

rleigh-codelibre opened the pull request:
Add minimal tiff tests to test pixels only.  Avoids any timing ambiguity with the OME-TIFF writer close method.

Commit:
Add separate OME-TIFF tests

dgault commented on src/main/java/ome/files/performance/OMETIFFPerformance.java:
```diff
@@ -0,0 +1,201 @@
+        Timepoint write_init = null;
+        Timepoint close_start = null;
+
+          {
```
Are the new braces here for readability? Also some of the indentations appear to be out of sync

rleigh-codelibre replied:
They are for limiting the scope of variables used for each phase of the test.  Commit pushed to reindent.

dgault commented:
Ive prove read through the code changes for the cpp, jace and java. Other than some minor formatting the actual structure of the tests alls looks to be good and the code appears to be fine.

Only remaining thing is to test if the new tests run as expected.

Commit:
java: Correct indentation

sbesson commented:
One thought that might be confirmed by the benchmarking run started earlier today. While the new approach might work for the BBBC and MitoCheck datasets (which are single OME-TIFF), I expect the `MinimalTiffReader` used in the new `pixels-performance` test will not work as expected against the multi-TIFF `tubhiswt` dataset making the metrics not directly comparable between both tests. Is that the expectation?

sbesson commented:
See https://ci.openmicroscopy.org/job/OME-FILES-performance-win/138/ for the Windows results with the latest state of this PR. Having briefly looked  at the tabular results for MitoCheck (where the pixeldata rates are measured to be the most extreme), I do not see any obvious differences in the values of `pixeldata.read` vs `ometiffdata.read.pixels` and `pixeldata.write` vs `ometiff.write.pixels` in both languages. This might suggests our current approximation of using `ometiffdata.{write,read}.pixels` as a proxy for the pixeldata reading/writing is valid. This might require confirmation via some statistical analysis.

rleigh-codelibre commented:
Some stats.  `p` is pixeldata (tiffwriter), `o` is ometiff (ometiffwriter).  MitoCheck pixel write total was not different to ometiff pixel write.  This was not the case for BBBC however.  tubhiswt not comparable due to only reading one series.  I'd conclude that it's not properly comparable due to the significant differences.

```
> p <- read.dataset(c("tubhiswt", "bbbc", "mitocheck"), "pixeldata", FALSE, TRUE)
> o <- read.dataset(c("tubhiswt", "bbbc", "mitocheck"), "ometiffdata", FALSE, TRUE)
> 
> pwrite <- subset(p, dataset=='MitoCheck' & Language=='C++' & Test=='write')$proc.real
> pwriteini <- subset(p, dataset=='MitoCheck' & Language=='C++' & Test=='write.init')$proc.real
> pwritepix <- subset(p, dataset=='MitoCheck' & Language=='C++' & Test=='write.pixels')$proc.real
> pwriteclo <- subset(p, dataset=='MitoCheck' & Language=='C++' & Test=='write.close')$proc.real
> owrite <- subset(o, dataset=='MitoCheck' & Language=='C++' & Test=='write')$proc.real
> owriteini <- subset(o, dataset=='MitoCheck' & Language=='C++' & Test=='write.init')$proc.real
> owritepix <- subset(o, dataset=='MitoCheck' & Language=='C++' & Test=='write.pixels')$proc.real
> owriteclo <- subset(o, dataset=='MitoCheck' & Language=='C++' & Test=='write.close')$proc.real
> 
> pwrite
 [1] 133 132 133 133 131 133 134 134 133 133 133 133 133 132 124 123 123 122 123
[20] 120
> owrite
 [1] 4493 4500 4506 4523 4480 4497 4523 4508 4491 4519 4489 4512 4504 4591 4573
[16] 4576 4488 4607 4486 4499
> pwriteini
 [1] 0 0 0 0 1 1 1 0 1 1 0 1 0 1 0 0 0 0 0 0
> owriteini
 [1] 1800 1803 1798 1799 1797 1803 1801 1804 1801 1798 1803 1804 1804 1826 1826
[16] 1825 1798 1804 1798 1800
> pwritepix
 [1] 132 132 133 132 130 132 132 133 132 132 132 132 133 130 123 123 123 122 122
[20] 120
> owritepix
 [1] 116 117 119 119 119 120 122 122 125 125 125 127 127 137 133 134 126 127 127
[20] 128
> pwriteclo
 [1] 1 0 0 1 0 0 1 1 0 0 1 0 0 1 1 0 0 0 1 0
> owriteclo
 [1] 2577 2580 2589 2605 2564 2574 2600 2582 2565 2596 2561 2581 2573 2628 2614
[16] 2617 2564 2676 2561 2571
> 
> mean (pwrite)
[1] 129.75
> mean (owrite)
[1] 4518.25
> mean (pwriteini)
[1] 0.35
> mean (owriteini)
[1] 1804.6
> mean (pwritepix)
[1] 129
> mean (owritepix)
[1] 124.75
> mean (pwriteclo)
[1] 0.4
> mean (owriteclo)
[1] 2588.9
> 
> t.test(pwritepix, owritepix)

        Welch Two Sample t-test

data:  pwritepix and owritepix
t = 2.5843, df = 36.722, p-value = 0.01387
alternative hypothesis: true difference in means is not equal to 0
95 percent confidence interval:
 0.9169354 7.5830646
sample estimates:
mean of x mean of y 
   129.00    124.75 

> t.test(pwrite, pwritepix)

        Welch Two Sample t-test

data:  pwrite and pwritepix
t = 0.49124, df = 37.88, p-value = 0.6261
alternative hypothesis: true difference in means is not equal to 0
95 percent confidence interval:
 -2.341035  3.841035
sample estimates:
mean of x mean of y 
   129.75    129.00 

> t.test(pwrite, owritepix)

        Welch Two Sample t-test

data:  pwrite and owritepix
t = 2.9693, df = 37.351, p-value = 0.00519
alternative hypothesis: true difference in means is not equal to 0
95 percent confidence interval:
 1.589173 8.410827
sample estimates:
mean of x mean of y 
   129.75    124.75 

> 
> pwrite <- subset(p, dataset=='BBBC' & Language=='C++' & Test=='write')$proc.real
> pwriteini <- subset(p, dataset=='BBBC' & Language=='C++' & Test=='write.init')$proc.real
> pwritepix <- subset(p, dataset=='BBBC' & Language=='C++' & Test=='write.pixels')$proc.real
> pwriteclo <- subset(p, dataset=='BBBC' & Language=='C++' & Test=='write.close')$proc.real
> owrite <- subset(o, dataset=='BBBC' & Language=='C++' & Test=='write')$proc.real
> owriteini <- subset(o, dataset=='BBBC' & Language=='C++' & Test=='write.init')$proc.real
> owritepix <- subset(o, dataset=='BBBC' & Language=='C++' & Test=='write.pixels')$proc.real
> owriteclo <- subset(o, dataset=='BBBC' & Language=='C++' & Test=='write.close')$proc.real
> 
> pwrite
 [1]  81637  82986  83642  85052  86715  87911  89914  91081  97897  96283
[11] 108084 111438 106389 109778 114147 114495 119089 122766 126885 128789
> owrite
 [1]  85932  87416  90062  92740  95763  98819 100785 103679 106615 109488
[11] 113596 117587 120245 125418 126129 125765 127165 127945 128456 129515
> pwriteini
 [1] 0 1 1 1 1 0 1 1 1 0 0 0 0 1 0 0 1 0 1 0
> owriteini
 [1] 1705 1714 1717 1720 1722 1733 1736 1745 1752 1762 1772 1773 1786 1852 1838
[16] 1807 1795 1798 1811 1814
> pwritepix
 [1]  81615  82962  83618  85029  86691  87888  89890  91058  97873  96259
[11] 108060 111415 106366 109754 114124 114472 119064 122742 126860 128765
> owritepix
 [1]  82753  84230  86818  89519  92481  95538  97461 100348 103267 106113
[11] 110048 114176 116793 121865 122391 122233 123667 124442 124945 126002
> pwriteclo
 [1] 22 23 23 22 23 23 23 22 23 24 24 23 23 23 23 23 24 24 24 24
> owriteclo
 [1] 1474 1472 1527 1501 1560 1548 1588 1586 1596 1613 1776 1638 1666 1701 1900
[16] 1725 1703 1705 1700 1699
> 
> mean (pwrite)
[1] 102248.9
> mean (owrite)
[1] 110656
> mean (pwriteini)
[1] 0.5
> mean (owriteini)
[1] 1767.6
> mean (pwritepix)
[1] 102225.2
> mean (owritepix)
[1] 107254.5
> mean (pwriteclo)
[1] 23.15
> mean (owriteclo)
[1] 1633.9
> 
> t.test(pwritepix, owritepix)

        Welch Two Sample t-test

data:  pwritepix and owritepix
t = -1.0256, df = 37.961, p-value = 0.3116
alternative hypothesis: true difference in means is not equal to 0
95 percent confidence interval:
 -14956.561   4898.061
sample estimates:
mean of x mean of y 
 102225.2  107254.5 

> t.test(pwrite, pwritepix)

        Welch Two Sample t-test

data:  pwrite and pwritepix
t = 0.0047471, df = 38, p-value = 0.9962
alternative hypothesis: true difference in means is not equal to 0
95 percent confidence interval:
 -10061.77  10109.07
sample estimates:
mean of x mean of y 
 102248.9  102225.2 

> t.test(pwrite, owritepix)

        Welch Two Sample t-test

data:  pwrite and owritepix
t = -1.0208, df = 37.961, p-value = 0.3138
alternative hypothesis: true difference in means is not equal to 0
95 percent confidence interval:
 -14933.007   4921.807
sample estimates:
mean of x mean of y 
 102248.9  107254.5
```

sbesson commented:
Discussed the last results with @rleigh-codelibre. The analysis above suggests our current `pixeldata.{read/write}` values are slightly underestimating a true pixeldata-only case. It looks like the difference is marginal (5% at most) and consistent across datasets/implementations. This should not affect the relative performance measures and be a minimal correction to the current measurements.

Also this refinement does not explain the the surprisingly high pixeldata rates for the MitoCheck sample which might be related to the small size of the sample nor the increase of the BBBC timings which seem to affect only Windows. Thus there is no point in including this new benchmark test for the first version of this benchmarking.

Once this repository is tagged with `0.1.0` we can certainly come back to these enhancements and investigate the two points raised above.

Commit:
Pixeldata tests can read filesets

Commit:
Add scripts to run Java tests

- Use maven dependency plugin to get jars
- Scripts set classpath
- Remove unnecessary maven profiles

rleigh-codelibre commented:
@sbesson pixeldata tests now read in whole filesets

Commit:
scripts: Glob tubhiswt TIFFs for pixel performance tests

Commit:
scripts: Remove exec from batch file

Commit:
scripts: Add  eqivalent for batch files

Commit:
scripts: Add batch debug message

Commit:
scripts: Remove maven goal

rleigh-codelibre commented:
See https://ci.openmicroscopy.org/view/Files/job/OME-FILES-performance-win/147/

sbesson commented on scripts/run_benchmarking:
```diff
@@ -19,25 +19,33 @@ for test in bbbc mitocheck tubhiswt; do
         ;;
         tubhiswt)
             input=${datapath}/tubhiswt-4D/tubhiswt_C0_TP0.ome.tif
+            input=${datapath}/tubhiswt-4D/*.tif
```
Is this supposed to be `inputglob`

sbesson reviewed (commented):
In https://ci.openmicroscopy.org/view/Experimental/job/OME-FILES-performance-win/147/consoleText, the glob logic  is causing

```
Running java -classpath "target/classes;target/dependency/*" ome.files.performance.PixelsPerformance 1 "tubhiswt_C0_TP0.ome.tif" "tubhiswt_C0_TP1.ome.tif" "tubhiswt_C0_TP10.ome.tif" "tubhiswt_C0_TP11.ome.tif" "tubhiswt_C0_TP12.ome.tif" "tubhiswt_C0_TP13.ome.tif" "tubhiswt_C0_TP14.ome.tif" "tubhiswt_C0_TP15.ome.tif" "tubhiswt_C0_TP16.ome.tif" "tubhiswt_C0_TP17.ome.tif" "tubhiswt_C0_TP18.ome.tif" "tubhiswt_C0_TP19.ome.tif" "tubhiswt_C0_TP2.ome.tif" "tubhiswt_C0_TP20.ome.tif" "tubhiswt_C0_TP21.ome.tif" "tubhiswt_C0_TP22.ome.tif" "tubhiswt_C0_TP23.ome.tif" "tubhiswt_C0_TP24.ome.tif" "tubhiswt_C0_TP25.ome.tif" "tubhiswt_C0_TP26.ome.tif" "tubhiswt_C0_TP27.ome.tif" "tubhiswt_C0_TP28.ome.tif" "tubhiswt_C0_TP29.ome.tif" "tubhiswt_C0_TP3.ome.tif" "tubhiswt_C0_TP30.ome.tif" "tubhiswt_C0_TP31.ome.tif" "tubhiswt_C0_TP32.ome.tif" "tubhiswt_C0_TP33.ome.tif" "tubhiswt_C0_TP34.ome.tif" "tubhiswt_C0_TP35.ome.tif" "tubhiswt_C0_TP36.ome.tif" "tubhiswt_C0_TP37.ome.tif" "tubhiswt_C0_TP38.ome.tif" "tubhiswt_C0_TP39.ome.tif" "tubhiswt_C0_TP4.ome.tif" "tubhiswt_C0_TP40.ome.tif" "tubhiswt_C0_TP41.ome.tif" "tubhiswt_C0_TP42.ome.tif" "tubhiswt_C0_TP5.ome.tif" "tubhiswt_C0_TP6.ome.tif" "tubhiswt_C0_TP7.ome.tif" "tubhiswt_C0_TP8.ome.tif" "tubhiswt_C0_TP9.ome.tif" "tubhiswt_C1_TP0.ome.tif" "tubhiswt_C1_TP1.ome.tif" "tubhiswt_C1_TP10.ome.tif" "tubhiswt_C1_TP11.ome.tif" "tubhiswt_C1_TP12.ome.tif" "tubhiswt_C1_TP13.ome.tif" "tubhiswt_C1_TP14.ome.tif" "tubhiswt_C1_TP15.ome.tif" "tubhiswt_C1_TP16.ome.tif" "tubhiswt_C1_TP17.ome.tif" "tubhiswt_C1_TP18.ome.tif" "tubhiswt_C1_TP19.ome.tif" "tubhiswt_C1_TP2.ome.tif" "tubhiswt_C1_TP20.ome.tif" "tubhiswt_C1_TP21.ome.tif" "tubhiswt_C1_TP22.ome.tif" "tubhiswt_C1_TP23.ome.tif" "tubhiswt_C1_TP24.ome.tif" "tubhiswt_C1_TP25.ome.tif" "tubhiswt_C1_TP26.ome.tif" "tubhiswt_C1_TP27.ome.tif" "tubhiswt_C1_TP28.ome.tif" "tubhiswt_C1_TP29.ome.tif" "tubhiswt_C1_TP3.ome.tif" "tubhiswt_C1_TP30.ome.tif" "tubhiswt_C1_TP31.ome.tif" "tubhiswt_C1_TP32.ome.tif" "tubhiswt_C1_TP33.ome.tif" "tubhiswt_C1_TP34.ome.tif" "tubhiswt_C1_TP35.ome.tif" "tubhiswt_C1_TP36.ome.tif" "tubhiswt_C1_TP37.ome.tif" "tubhiswt_C1_TP38.ome.tif" "tubhiswt_C1_TP39.ome.tif" "tubhiswt_C1_TP4.ome.tif" "tubhiswt_C1_TP40.ome.tif" "tubhiswt_C1_TP41.ome.tif" "tubhiswt_C1_TP42.ome.tif" "tubhiswt_C1_TP5.ome.tif" "tubhiswt_C1_TP6.ome.tif" "tubhiswt_C1_TP7.ome.tif" "tubhiswt_C1_TP8.ome.tif" "tubhiswt_C1_TP9.ome.tif" D:\jenkins\workspace\OME-FILES-performance-win\out\tubhiswt-java.ome.tiff D:\jenkins\workspace\OME-FILES-performance-win\results\tubhiswt-pixeldata-win-java-15.tsv
pass 0: read init...Error: caught exception: java.io.FileNotFoundException: tubhiswt_C0_TP0.ome.tif (The system cannot find the file specified)
java.io.FileNotFoundException: tubhiswt_C0_TP0.ome.tif (The system cannot find the file specified)
	at java.io.RandomAccessFile.open(Native Method)
	at java.io.RandomAccessFile.<init>(RandomAccessFile.java:241)
	at loci.common.NIOFileHandle.<init>(NIOFileHandle.java:123)
	at loci.common.Location.getHandle(Location.java:320)
	at loci.common.RandomAccessInputStream.<init>(RandomAccessInputStream.java:131)
	at loci.formats.in.MinimalTiffReader.initFile(MinimalTiffReader.java:432)
	at loci.formats.FormatReader.setId(FormatReader.java:1399)
	at ome.files.performance.PixelsPerformance.main(PixelsPerformance.java:111)
```

Looking at the old and new pixels classes, the diff without whitespace is pretty minimal. If we are want to go down that route, is there a case for thinking of the proposal into a generic `PixelsPerformance` base class with concrete implementations e.g. `TiffPerformance`, `OMETiffPerformance`.

sbesson commented:
As shown by the last results suggest more work would be required to fix the multi-file case. Initial results suggested that the existing usage OME-TIFF readers/writers does not bias the relative comparison.

As we will be cleaning up and extending these benchmarks to cover tiling, closing this PR to prevent conflicts. We can revisit its content and cherry-pick some of the added functionalities as we make progress.
